### PR TITLE
update rollouts to re-evaluate enrollment if they are disqualified for targeting or bucketing

### PR DESCRIPTION
**Pull request by jeddai**
[EXP-3649](https://mozilla-hub.atlassian.net/browse/EXP-3649)

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/nimbus/src/enrollment.rs [hunk 1]
```diff
@@ -167,8 +169,8 @@ impl ExperimentEnrollment {
         targeting_helper: &NimbusTargetingHelper,
         out_enrollment_events: &mut Vec<EnrollmentChangeEvent>,
     ) -> Result<Self> {
-        Ok(match self.status {
-            EnrollmentStatus::NotEnrolled { .. } | EnrollmentStatus::Error { .. } => {
+        Ok(match (updated_experiment.is_rollout, self.status.clone()) {
```
Comment (eliserichards) [inline components/nimbus/src/enrollment.rs:172]: 👏

File: components/nimbus/src/enrollment.rs [hunk 2]
```diff
@@ -190,11 +192,14 @@ impl ExperimentEnrollment {
...
-            } => {
+            (
+                _,
+                EnrollmentStatus::Enrolled {
+                    ref branch,
+                    ref reason,
+                    ..
+                },
+            ) => {
                 if !is_user_participating {
```
Comment (eliserichards) [inline components/nimbus/src/enrollment.rs:203]: Could `!is_user_participating` be opt-out OR `NotSelected` now?

File: components/nimbus/src/enrollment.rs [hunk 3]
```diff
@@ -252,11 +259,32 @@ impl ExperimentEnrollment {
                     }
                 }
             }
-            EnrollmentStatus::Disqualified {
-                ref branch,
-                enrollment_id,
-                ..
-            } => {
+            (
+                true,
```
Comment (eliserichards) [inline components/nimbus/src/enrollment.rs:263]: Why is this `true` instead of `_` (like ln 196)?

File: components/nimbus/src/nimbus_client.rs [hunk 4]
```diff
@@ -325,7 +326,15 @@ impl NimbusClient {
...
-                EnrollmentStatus::WasEnrolled { .. } => {
+                EnrollmentStatus::WasEnrolled { .. }
+                | EnrollmentStatus::Disqualified {
+                    reason: DisqualifiedReason::NotSelected,
+                    ..
+                }
+                | EnrollmentStatus::Disqualified {
+                    reason: DisqualifiedReason::NotTargeted,
+                    ..
+                } => {
```
Comment (eliserichards) [inline components/nimbus/src/nimbus_client.rs:337]: Do we need the `Disqualified` `NotSelected`/`NotTargeted` here? Or will `WasEnrolled` cover those cases?

**Review by eliserichards — COMMENTED** [review]
Great tests! 🥇 This looks good, I just have a few ~possibly stupid~ questions:

File: components/nimbus/src/enrollment.rs [hunk 5]
```diff
@@ -270,6 +275,26 @@ impl ExperimentEnrollment {
                             branch: branch.clone(),
                         },
                     }
+                } else if updated_experiment.is_rollout
```
Comment (eliserichards) [inline components/nimbus/src/enrollment.rs:278]: gorgeous 👏

**Review by eliserichards — APPROVED** [review]
🚢

File: components/nimbus/src/tests/test_enrollment.rs [hunk 6]
```diff
@@ -877,7 +880,144 @@ fn test_rollout_unenrolls_when_bucketing_changes() -> Result<()> {
...
+    assert_eq!(enrollments.len(), 1);
+    let enr = enrollments.get(0).unwrap();
+    assert_eq!(&enr.slug, slug);
+    assert!(matches!(&enr.status, EnrollmentStatus::Enrolled { .. }));
+
+    Ok(())
+}
+
+#[test]
+fn test_experiment_does_not_reenroll_from_disqualified_not_selected_or_not_targeted() -> Result<()>
```
Comment (jaredlockhart) [inline components/nimbus/src/tests/test_enrollment.rs:972]: Just a small note about this:

Experiment SHOULD re-attempt enrollment if it previously attempted and did not match targeting

File: CHANGELOG.md [hunk 7]
```diff
@@ -26,6 +26,7 @@
 - Add `enrollments` value to `TargetingAttributes` — it is a set of strings containing all enrollments, past and present ([#5685](https://github.com/mozilla/application-services/pull/5685)).
   - _Note: This change only applies to stateful uses of the Nimbus SDK, e.g. mobile_
 - Add ability to enroll selected features multiple times (coenrollment) ([#5684](https://github.com/mozilla/application-services/pull/5684), [#5697](https://github.com/mozilla/application-services/pull/5697)).
+- Update enrollment logic to re-evaluate enrollment for rollouts ([#5716](https://github.com/mozilla/application-services/pull/5716)).
```
Comment (jhugman) [inline CHANGELOG.md:29]: Could you add a sentence and PR link to https://github.com/mozilla/application-services/pull/5716/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR24

File: components/nimbus/src/nimbus_client.rs [hunk 8]
```diff
@@ -325,7 +326,11 @@ impl NimbusClient {
                     is_enrolled_set.insert(ee.slug.clone());
                     all_enrolled_set.insert(ee.slug.clone());
                 }
-                EnrollmentStatus::WasEnrolled { .. } => {
+                EnrollmentStatus::WasEnrolled { .. }
+                | EnrollmentStatus::Disqualified {
+                    reason: DisqualifiedReason::NotSelected | DisqualifiedReason::NotTargeted,
+                    ..
+                } => {
```
Comment (jhugman) [inline components/nimbus/src/nimbus_client.rs:333]: If `all_enrolled_set` is the set of experiments that the user is or was ever in an enrolled state, then this should include `OptOut` and `Error` disqualifications, no?
```suggestion
                | EnrollmentStatus::Disqualified { .. } => {
```

File: components/nimbus/src/enrollment.rs [hunk 9]
```diff
@@ -270,6 +275,26 @@ impl ExperimentEnrollment {
...
+                    && matches!(
+                        self.status,
+                        EnrollmentStatus::Disqualified {
+                            reason: DisqualifiedReason::NotSelected
+                                | DisqualifiedReason::NotTargeted,
```
Comment (jhugman) [inline components/nimbus/src/enrollment.rs:283]: This is confusing; you're matching on `self.status` and you've already matched on `self.status` (to get into this arm of the match).

Consider adding `reason` to the [outer match arm](https://github.com/mozilla/application-services/pull/5716/files#diff-5be1ba032964b639be01987de468be59027bf9f057f8907b4c324a749d465080R260-R264), and matching on that.

File: components/nimbus/src/nimbus_client.rs (hunk 8 above)
Comment (jeddai) [reply components/nimbus/src/nimbus_client.rs:333]: Yeah that's fair, I'll adjust that

File: components/nimbus/src/enrollment.rs (hunk 9 above)
Comment (jeddai) [reply components/nimbus/src/enrollment.rs:283]: 🤦 Idk how I missed that, great callout

File: components/nimbus/src/nimbus_client.rs (hunk 8 above)
Comment (jeddai) [reply components/nimbus/src/nimbus_client.rs:333]: Also with this I'm adding a section to the unit test where it reads from the database to fetch all the enrollments and verifies that each of the experiments are in the status we expect them to be in. 

The choice to do this is two-fold, it validates that the logic surrounding disqualification for various reasons continues to work as expected but more importantly it serves as documentation of what's being tested.

**Pull request merged**